DOCS-11518-mqtt3-first-version-kt (#1579)
mulesoft/docs-connectors: mqtt3/1.0.0/modules/ROOT/pages/mqtt3-connector-listener.adoc

= Listen for New Messages with MQTT3 Connector

The Anypoint Connector for MQTT3 (MQTT3 Connector) *On New Message* (`<mqtt3:listener>`) source enables you to listen for new incoming messages using one or more topic filters (`topicFilter`). Each topic filter has a specific quality of service (`qos`) configured.

> In the Configuration XML editor, the `<mqtt3:listener>` and `<mqtt3:topic>` configurations look like this:

Reviewer: Maybe indicate that this is an example, either in the section heading or in this sentence: "In this example, the..." or "In these examples, the..."

[source,xml,linenums]
----
    <flow name="listenerAuthorQuotes">
        <mqtt3:listener config-ref="MQTT_Config">
            <mqtt3:topics>
                <mqtt3:topic topicFilter="quotes/terryPratchett" qos="EXACTLY_ONCE"/>
                <mqtt3:topic topicFilter="quotes/neilGaiman" qos="AT_LEAST_ONCE" />
                <mqtt3:topic topicFilter="quotes/ianMcEwan" qos="AT_MOST_ONCE" />
            </mqtt3:topics>
        </mqtt3:listener>
    </flow>
----

== Configure Single-Level Wildcards in Topic Filters

> Configure single-level wildcards to listen to multiple topics. Single-level wildcards (`+`) enable the subscriber to receive messages published to all topics matching a specific structure.

Reviewer: Suggestion: matching > that matches

In the following example, you configure the `topicFilter` parameter with the structure `quotes/+/authors`:

[source,xml,linenums]
----
    <flow name="quotes">
        <mqtt3:listener config-ref="${config}">
            <mqtt3:topics>
                <mqtt3:topic topicFilter="quotes/+/authors"/>
            </mqtt3:topics>
        </mqtt3:listener>

        <logger level="DEBUG" message="A quote has been published to #[attributes.topic]: #[payload]"/>
    </flow>
----

Messages published to the following topics trigger the listener:

* `quotes/british/authors`
* `quotes/american/authors`

> However, messages published to the following topics do not trigger the listener:

Reviewer: I really like that you give examples for what triggers the listener and what doesn't trigger the listener.

* `names/british/authors`
* `quotes/american/writers`

== Configure Multi-Level Wildcards in Topic Filters

Configure multi-level wildcards to enable the listener to subscribe to all topics that share the same root (everything that comes before the hash `#` symbol).

In the following example, you configure the `topicFilter` parameter with the structure `quotes/england/#`:

[source,xml,linenums]
----
    <flow name="listenerArgentinaTemperature">
        <mqtt3:listener config-ref="${config}">
            <mqtt3:topics>
                <mqtt3:topic topicFilter="quotes/england/#"/>
            </mqtt3:topics>
        </mqtt3:listener>

        <logger level="DEBUG" message="A quote has been published to #[attributes.topic]: #[payload]"/>
    </flow>
----

Messages published to the following topics trigger the listener:

* `quotes/england/authors/terryPratchett`
* `quotes/england/authors/neilGaiman`
* `quotes/england/actors`

However, messages published to the following topics do not trigger the listener:

* `quotes/american/actors`
* `phrases/england/authors/neilGaiman`

== Listeners Sharing a Configuration

> Listeners that share a configuration element also share a connection and client ID. This is important to keep in mind especially in the cases where one of the listeners is stopped but the other isn't.

Reviewer: keep in mind, especially when one...

Less wordy: This is important to remember, especially...

> Suppose we have the following configuration element and listeners in our application:

Reviewer: This sentence is not needed because you introduce the XML in the next paragraph.

    <mqtt3:connection clientId="smart-bentley-123" url="tcp://127.0.0.1:1883" />
</mqtt3:config>
<flow name="listenerReaderA">
   <mqtt3:listener config-ref="MQTT_Config">
            <mqtt3:topics>
                <mqtt3:topic topicFilter="shakespeare"/>
                <mqtt3:topic topicFilter="terryPratchett"/>
            </mqtt3:topics>
   </mqtt3:listener>
   <logger level="INFO"  message="Received message '#[payload]' with at topic #[attributes.topic] with qos #[attributes.qos]">
</flow>
<flow name="listenerReaderB">
    <mqtt3:listener config-ref="MQTT_Config">
            <mqtt3:topics>
                <mqtt3:topic topicFilter="neilGaiman"/>
                <mqtt3:topic topicFilter="terryPratchett"/>
            </mqtt3:topics>
   </mqtt3:listener>
   <logger level="INFO"  message="Received message '#[payload]' with at topic #[attributes.topic] with qos #[attributes.qos]">
</flow>
----

As you can see from the XML example, the listeners also share a subscription to the topic `terryPratchett`. Whichever listener subscribes first to the topic can set the subscription quality of service (QoS). Only one subscription with one quality of service level can exist for that topic.

> 

Reviewer: I suggest removing  "The second observation is that" and start with "If the...." It makes the text very wording and is not needed.

> The second observation is that, if the `listenerReaderB` flow stops, the `listenerReaderA` flow still receives and processes the messages for the `terryPratchett` topic.  Even if the configuration element specifies a `cleanSession=false`, the messages that were processed while `listenerReaderB` was offline, are not resent to this listener.

Reviewer: I suggest removing "The second observation is that" and start with "If the...." It makes the text very wording and is not needed.

Reviewer: ```suggestion
If the `listenerReaderB` flow stops, the `listenerReaderA` flow still receives and processes the messages for the `terryPratchett` topic.  Even if the configuration element specifies `cleanSession=false`, the messages processed while `listenerReaderB` was offline, are not resent to `listenerReaderB`.
```

= Listen for New Messages with MQTT3 Connector

> The MQTT3 Connector allows you to listen for new incoming messages for one or more topic filters and with a specific quality of service for each one.

Reviewer 2: ```suggestion
MQTT3 Connector enables you to listen for new incoming messages for one or more topic filters and with a specific quality of service for each one.
```

The listener could subscribe to multiple individual topics:

[source,xml,linenums]
----
    <flow name="listenerAuthorQuotes">
        <mqtt3:listener config-ref="MQTT_Config">
            <mqtt3:topics>
                <mqtt3:topic topicFilter="quotes/terryPratchett" qos="EXACTLY_ONCE"/>
                <mqtt3:topic topicFilter="quotes/neilGaiman" qos="AT_LEAST_ONCE" />
                <mqtt3:topic topicFilter="quotes/ianMcEwan" qos="AT_MOST_ONCE" />
            </mqtt3:topics>
        </mqtt3:listener>
    </flow>
----

== Single-Level WildCards

You can use single-level wildcards to listen to multiple topics:

[source,xml,linenums]
----
    <flow name="listenerArgentinaTemperature">
        <mqtt3:listener config-ref="${config}">
            <mqtt3:topics>
                <mqtt3:topic topicFilter="quotes/+/authors"/>
            </mqtt3:topics>
        </mqtt3:listener>

        <logger level="DEBUG" message="A quote has been published to #[attributes.topic]: #[payload]"/>
    </flow>
----

> This will enable the subscriber to receive messages published to all topics matching the structure 'quotes/+/authors'.

Reviewer 2: ```suggestion
This enables the subscriber to receive messages published to all topics matching the structure `quotes/+/authors`.
```

> So, messages published to the following topics would reach trigger our message listener:

Reviewer 2: ```suggestion
So, messages published to the following topics would trigger the message listener:
```

* quotes/british/authors
* quotes/american/authors

> However, messages published to the following topics would NOT trigger our message listener:

Reviewer 2: ```suggestion
However, messages published to the following topics would NOT trigger the message listener:
```

* names/british/authors
* quotes/american/writers

== Multi-Level WildCards

> Or multi-level wildcards:

Reviewer 2: Probably should have a full lead-in sentence for this section.

[source,xml,linenums]
----
    <flow name="listenerArgentinaTemperature">
        <mqtt3:listener config-ref="${config}">
            <mqtt3:topics>
                <mqtt3:topic topicFilter="quotes/england/#"/>
            </mqtt3:topics>
        </mqtt3:listener>

        <logger level="DEBUG" message="A quote has been published to #[attributes.topic]: #[payload]"/>
    </flow>
----

> Multi-level wildcards allow the listener to subscribe to all topics that share same root (everything that comes before '#').

Reviewer 2: ```suggestion
Multi-level wildcards allow the listener to subscribe to all topics that share the same root (everything that comes before '#').
```

> So, messages published to the following topics would reach trigger our message listener:

Reviewer 2: ```suggestion
So, messages published to the following topics would trigger the message listener:
```

* quotes/england/authors/terryPratchett
* quotes/england/authors/neilGaiman
* quotes/england/actors

> However, messages published to the following topics would NOT trigger our message listener:

Reviewer 2: ```suggestion
However, messages published to the following topics would NOT trigger the message listener:
```

= Listen for New Messages with MQTT3 Connector

> The Anypoint Connector for MQTT3 (MQTT3 Connector) *On New Message* (`<mqtt3:listener>`) source enables you to listen for new incoming messages at one or more topic filters (`topicFilter`). Each topic filter has a specific quality of service (`qos`) configured.

Reviewer 2: ```suggestion
The Anypoint Connector for MQTT3 (MQTT3 Connector) *On New Message* (`<mqtt3:listener>`) source enables you to listen for new incoming messages using one or more topic filters (`topicFilter`). Each topic filter has a specific quality of service (`qos`) configured.
```

In the Configuration XML editor, the `<mqtt3:listener>` and `<mqtt3:topic>` configurations look like this:

[source,xml,linenums]
----
    <flow name="listenerAuthorQuotes">
        <mqtt3:listener config-ref="MQTT_Config">
            <mqtt3:topics>
                <mqtt3:topic topicFilter="quotes/terryPratchett" qos="EXACTLY_ONCE"/>
                <mqtt3:topic topicFilter="quotes/neilGaiman" qos="AT_LEAST_ONCE" />
                <mqtt3:topic topicFilter="quotes/ianMcEwan" qos="AT_MOST_ONCE" />
            </mqtt3:topics>
        </mqtt3:listener>
    </flow>
----

== Configure Single-Level Wildcards in Topic Filters

> Configure single-level wildcards to listen to multiple topics. Single-level wildcards (`+`) enables the subscriber to receive messages published to all topics matching a specific structure.

Reviewer 2: ```suggestion
Configure single-level wildcards to listen to multiple topics. Single-level wildcards (`+`) enable the subscriber to receive messages published to all topics matching a specific structure.
```

In the following example, you configure the `topicFilter` parameter with the structure `quotes/+/authors`:

[source,xml,linenums]
----
    <flow name="quotes">
        <mqtt3:listener config-ref="${config}">
            <mqtt3:topics>
                <mqtt3:topic topicFilter="quotes/+/authors"/>
            </mqtt3:topics>
        </mqtt3:listener>

        <logger level="DEBUG" message="A quote has been published to #[attributes.topic]: #[payload]"/>
    </flow>
----

Messages published to the following topics trigger the listener:

* `quotes/british/authors`
* `quotes/american/authors`

However, messages published to the following topics do not trigger the listener:

* `names/british/authors`
* `quotes/american/writers`

== Configure Multi-Level Wildcards in Topic Filters

> Configure multi-level wildcards to enable the listener to subscribe to all topics that share the same root, everything that comes before the hash `#` symbol.

Reviewer 2: ```suggestion
Configure multi-level wildcards to enable the listener to subscribe to all topics that share the same root (everything that comes before the hash `#` symbol).
```
